include/lbann/utils/dnn_lib/cudnn/dropout.hpp

```diff
@@ -0,0 +1,123 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+#ifndef LBANN_UTILS_DNN_LIB_CUDNN_DROPOUT_HPP_
+#define LBANN_UTILS_DNN_LIB_CUDNN_DROPOUT_HPP_
+
+#include "lbann/utils/cudnn.hpp"
+#include "lbann/utils/gpu/helpers.hpp"
+
+#include "utils.hpp"
+
+namespace lbann
+{
+
+namespace cudnn
+{
+
+template <typename TensorDataType>
+void get_dropout_state_size(size_t& size,
+                            El::Matrix<TensorDataType, El::Device::GPU> m_states)
+{
+  auto handle_manager = internal::make_default_handle_manager(gpu::get_sync_info(m_states));
+  CHECK_CUDNN(cudnnDropoutGetStatesSize(handle_manager.get(), &size));
+}
+
+template <typename TensorDataType>
+void dropout_forward(DropoutDescriptor dropoutDesc,
+                     TensorDescriptor xDesc,
+                     El::AbstractMatrix<TensorDataType> const& x,
+                     TensorDescriptor yDesc,
+                     El::AbstractMatrix<TensorDataType>& y,
+                     El::AbstractMatrix<TensorDataType>& workspace,
+                     El::SyncInfo<El::Device::GPU> const& si)
+{
+  size_t size;
+  CHECK_CUDNN(cudnnDropoutGetReserveSpaceSize(xDesc, &size));
+  workspace.Resize((size + sizeof(TensorDataType) - 1) / sizeof(TensorDataType), 1);
```

Comment:
IMO, since the workspace is managed by the layer, we should leave this in the calling code. I recommend exposing a `size_t get_dropout_reserve_space_size();` function, similar to `get_dropout_state_size()`.

Reply:
An additional thought down this same path is the following: since the reserve space is assumed not to change between forward prop and backward prop, it might be best to leave all explicit modification of its structure in the calling code. Even though this is likely correct, I think it's best to leave that burden with the layer implementer as there's a contract they need to guarantee in the layer itself.

Reply:
Addressed in 179674383264458eb7765595c243ebdafc290e6a